### PR TITLE
Revise create container process and trace it

### PR DESCRIPTION
**Pull request by CMGS**

### CODE REVIEW

File: store/etcd/processing.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+func (k *Krypton) loadProcessing(ctx context.Context, opts *types.DeployOptions, nodesInfo []types.NodeInfo) ([]types.NodeInfo, error) {
+	processingKey := filepath.Join(containerProcessingPrefix, opts.Name, opts.Entrypoint.Name)
+	resp, err := k.etcd.Get(ctx, processingKey, &etcdclient.GetOptions{Recursive: true})
+	if err != nil {
+		if etcdclient.IsKeyNotFound(err) {
+			err = nil
+		}
+		return nodesInfo, nil
```
Comment (ScienJus) [inline store/etcd/processing.go]: `return nodesInfo, err`

File: cluster/calcium/create_container.go [hunk 2]
```diff
@@ -95,100 +97,34 @@ func (c *Calcium) createContainerWithMemoryPrior(ctx context.Context, opts *type
...
 
-func (c *Calcium) doCreateContainerWithMemoryPrior(ctx context.Context, nodeInfo types.NodeInfo, opts *types.DeployOptions, index int) []*types.CreateContainerMessage {
+func (c *Calcium) doCreateContainer(ctx context.Context, nodeInfo types.NodeInfo, opts *types.DeployOptions, index int) []*types.CreateContainerMessage {
 	ms := make([]*types.CreateContainerMessage, nodeInfo.Deploy)
 
 	node, err := c.getAndPrepareNode(ctx, opts.Podname, nodeInfo.Name, opts.Image)
 	if err != nil {
-		log.Errorf("[doCreateContainerWithMemoryPrior] Get and prepare node error %v", err)
+		log.Errorf("[doCreateContainer] Get and prepare node error %v", err)
 		for i := 0; i < nodeInfo.Deploy; i++ {
```
Comment (wrfly) [inline cluster/calcium/create_container.go:106]: Can we just range the `nodeInfo.CPUPlan` here? If `nodeInfo.CPUPlan` is empty, just fill the message's error.

File: cluster/calcium/helper.go [hunk 3]
```diff
@@ -226,6 +226,7 @@ func pullImage(ctx context.Context, node *types.Node, image, auth string) error
 		return nil
 	}
 
+	log.Infof("[pullImage] Image not cached, pulling")
```
Comment (wrfly) [inline cluster/calcium/helper.go]: Have no idea why `Info` here :confused:

File: cluster/calcium/resource.go [hunk 4]
```diff
@@ -4,6 +4,8 @@ import (
 	"context"
 	"sort"
 
+	"github.com/sanity-io/litter"
+
```
Comment (wrfly) [inline cluster/calcium/resource.go:8]: Maybe we can remove this empty line, it breaks the style.

File: cluster/calcium/resource.go [hunk 5]
```diff
@@ -85,8 +87,23 @@ func (c *Calcium) allocResource(ctx context.Context, opts *types.DeployOptions,
...
+			s := litter.Sdump(nodeInfo.CPUPlan)
+			log.Infof("[allocResource] deploy %d to %s \n%s", nodeInfo.Deploy, nodeInfo.Name, s)
+		}
+	}()
+	return nodesInfo, c.bindProcessStatus(ctx, opts, nodesInfo)
+}
+
+func (c *Calcium) bindProcessStatus(ctx context.Context, opts *types.DeployOptions, nodesInfo []types.NodeInfo) error {
+	for _, nodeInfo := range nodesInfo {
+		if err := c.store.SaveProcessing(ctx, opts, nodeInfo); err != nil {
```
Comment (wrfly) [inline cluster/calcium/resource.go:102]: One fail, all fail? But the `doCreateContainer` simply `continue` to next one. Can be configurable here?

File: store/etcd/deploy.go [hunk 6]
```diff
@@ -4,20 +4,25 @@ import (
...
+// MakeDeployStatus get deploy status from store
 func (k *Krypton) MakeDeployStatus(ctx context.Context, opts *types.DeployOptions, nodesInfo []types.NodeInfo) ([]types.NodeInfo, error) {
-	key := fmt.Sprintf(containerDeployStatusKey, opts.Name, opts.Entrypoint.Name)
+	key := filepath.Join(containerDeployPrefix, opts.Name, opts.Entrypoint.Name)
 	resp, err := k.etcd.Get(ctx, key, &etcdclient.GetOptions{Recursive: true})
-	if err != nil && etcdclient.IsKeyNotFound(err) {
-		return k.doMakeDeployStatus(ctx, opts, nodesInfo)
+	if err == nil {
+		nodesInfo, err = k.doGetDeployStatus(ctx, resp, nodesInfo)
+	} else if err != nil && etcdclient.IsKeyNotFound(err) {
```
Comment (wrfly) [inline store/etcd/deploy.go]: `if err != nil && etcdclient.IsKeyNotFound(err)` == `if etcdclient.IsKeyNotFound(err)`

File: store/etcd/krypton.go [hunk 7]
```diff
@@ -26,11 +26,11 @@ const (
 	nodeContainers    = "/node/%s/containers"
 	nodeContainersKey = "/node/%s/containers/%s"
 
-	allContainersKey         = "/containers"
-	containerInfoKey         = "/containers/%s"
-	containerDeployPrefix    = "/deploy"
-	containerDeployStatusKey = "/deploy/%s/%s"
-	containerDeployKey       = "/deploy/%s/%s/%s/%s"
```
Comment (wrfly) [inline store/etcd/krypton.go:33]: Yeah, this way is stupid, maybe I wrote it somehow...

File: store/etcd/processing.go [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+	etcdclient "github.com/coreos/etcd/client"
+	"github.com/projecteru2/core/types"
+	log "github.com/sirupsen/logrus"
+)
+
+// SaveProcessing save processing status in etcd
+func (k *Krypton) SaveProcessing(ctx context.Context, opts *types.DeployOptions, nodeInfo types.NodeInfo) error {
+	///Appname/Entrypoint/Nodename/Ident
+	processingKey := filepath.Join(containerProcessingPrefix, opts.Name, opts.Entrypoint.Name, nodeInfo.Name, opts.ProcessIdent)
```
Comment (wrfly) [inline store/etcd/processing.go:18]: I usually break the line if it's too long, but it's OK.

File: store/etcd/processing.go [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+func (k *Krypton) loadProcessing(ctx context.Context, opts *types.DeployOptions, nodesInfo []types.NodeInfo) ([]types.NodeInfo, error) {
+	processingKey := filepath.Join(containerProcessingPrefix, opts.Name, opts.Entrypoint.Name)
+	resp, err := k.etcd.Get(ctx, processingKey, &etcdclient.GetOptions{Recursive: true})
+	if err != nil {
+		if etcdclient.IsKeyNotFound(err) {
+			err = nil
+		}
+		return nodesInfo, err
```
Comment (wrfly) [inline store/etcd/processing.go:46]: When `KeyNotFound`, is it really OK to return `nil` ? I'm not clear about this function.

File: cluster/calcium/create_container.go (hunk 2 above)
Comment (CMGS) [reply cluster/calcium/create_container.go:106]: In mem mode, CPUPlan alwasy empty

File: cluster/calcium/resource.go (hunk 4 above)
Comment (CMGS) [reply cluster/calcium/resource.go:8]: format by go format ... no idea why

File: cluster/calcium/resource.go (hunk 5 above)
Comment (CMGS) [reply cluster/calcium/resource.go:102]: This only affects the primary distribution and will be recalculated in next deploying. Fast fail will be better.

**Pull request closed without merging**